### PR TITLE
Submit blog post about tracetesting the OTel demo

### PR DESCRIPTION
**Pull request by adnanrahic**
Hi all!

My colleague @danielbdias wrote this piece based on his recently merged `tracetesting` contribution to the OTel demo that uses [Tracetest](https://github.com/kubeshop/tracetest). I only helped with reviews and formatting. Daniel is out on holiday this week, and he asked me to submit a PR on his behalf.

We thought that it would be nice to blog about it and contribute it to the OTel blog.
Would love to hear your thoughts about it! Do you think it makes sense? 

Thanks!

References:

- PR: https://github.com/open-telemetry/opentelemetry-demo/pull/877
- Original Issue: https://github.com/open-telemetry/opentelemetry-demo/issues/873

---

**Preview**: https://deploy-preview-3053--opentelemetry.netlify.app/blog/2023/testing-otel-demo/

### CODE REVIEW

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+body_class: otel-with-contributions-from
+spelling: cSpell:ignore apiserver choren tracetest kube kubelet containerd
+spelling: cSpell:ignore Adnan Rahić Hamric kubetracing healthcheck apparmor
+spelling: cSpell:ignore tracetesting oteldemo nanos
+---
+
+With contributions from [Adnan Rahić](https://github.com/adnanrahic) and
+[Ken Hamric](https://github.com/kdhamric).
+
+The [OpenTelemetry Demo](../../../ecosystem/demo/) is a system that simulates a
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:15]:
```suggestion
The [OpenTelemetry Demo](/ecosystem/demo/) is a system that simulates a
```

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+Telescope Shop, consisting of multiple microservices written in different
+languages, each handling a specific capability of this distributed system. Its
+purpose is to demonstrate how [OpenTelemetry](/) tools and SDKs can be used in
+an application to obtain telemetry for monitoring results and even to
+[track problems](../../../docs/demo/#scenarios) across multiple services.
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:20]:
```suggestion
[track problems](/docs/demo/#scenarios) across multiple services.
```

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+demo. It discusses the challenges we faced while building it, the outcomes, and
+how it can help the OpenTelemetry community test the demo and add features with
+greater confidence.
+
+## What is a trace-based test?
+
+Trace-based testing is a type of testing that validates a system's behavior by
+triggering an operation against the system and validating its result by checking
+the system output with the emitted
+[traces](../../../docs/concepts/signals/traces/) generated by the system during
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:40]:
```suggestion
[traces](/docs/concepts/signals/traces/) generated by the system during
```

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+4. Validate the operation output and the trace data by checking it against the
+   expected results. This involves analyzing the trace data to ensure the system
+   behaved as expected and the output was correct.
+5. If the trace data does not match the expected results, the test should fail.
+   With the trace data in hand, the developer can investigate the issue and make
+   necessary changes to the system or the test.
+
+![Example of Trace-based test running against a distributed system](./trace-based-testing-diagram.png)
+
+Example of Trace-based test running against a distributed system
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:65]: is this intended to be an image description? I think docsy has some capabilities for that.

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+Our testing mainly focused on the services involved in the main workflow of the
+system, which includes:
+
+1. A user navigating to the shop;
+2. Selecting a product;
+3. Deciding to make a purchase;
+4. And, completing the checkout process.
+
+We structured two types of tests based on
+[tests that currently exist in the demo](../../../docs/demo/tests/):
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:85]:
```suggestion
[tests that currently exist in the demo](/docs/demo/tests/):
```

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+call to the gRPC service with the method `oteldemo.CurrencyService.Convert` and
+a given payload.
+
+After that, in the `specs` section, we define which assertions to make against
+the trace and the operation result.
+
+We can see two types of assertions:
+
+- The first assertion is against a trace
+  [span](../../../docs/concepts/signals/traces/#spans) that the
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:153]:
```suggestion
  [span](/docs/concepts/signals/traces/#spans) that the
```

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+  output, where we check if the response body has an attribute `nanos` with a
+  value less or equal to `599380800`.
+
+### End-to-end tests
+
+The end-to-end tests are based on front-end tests using
+[Cypress](https://www.cypress.io/). We call the services through the API used by
+the front end and check if the service interaction between them is correct. We
+also verify if a trace is
+[propagated](../../../docs/concepts/signals/traces/#context-propagation)
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:169]:
```suggestion
[propagated](/docs/concepts/signals/traces/#context-propagation)
```

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+            "creditCardNumber": "4432-8015-6152-0454"
+          }
+        }
+  specs:
+  - name: "The frontend has been called with success"
+    selector: span[name="Test trigger"]
+    assertions:
+    - attr:response.status = 200
+  - selector: span[name="oteldemo.CheckoutService/PlaceOrder" rpc.system="grpc"
+			rpc.method="PlaceOrder" rpc.service="oteldemo.CheckoutService"]
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:282]:
```suggestion
            rpc.method="PlaceOrder" rpc.service="oteldemo.CheckoutService"]
```

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+    name: "The order was placed"
+    assertions:
+    - attr:app.user.id = "2491f868-88f1-4345-8836-d5d8511a9f83"
+    - attr:app.order.items.count = 1
+  - selector: span[name="oteldemo.PaymentService/Charge" rpc.system="grpc"
+			rpc.method="Charge" rpc.service="oteldemo.PaymentService"]
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:288]:
```suggestion
            rpc.method="Charge" rpc.service="oteldemo.PaymentService"]
```

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+    name: "The user was charged"
+    assertions:
+    - attr:rpc.grpc.status_code  =  0
+    - attr:selected_spans.count >= 1
+  - selector: span[name="oteldemo.ShippingService/ShipOrder" rpc.system="grpc"
+			rpc.method="ShipOrder" rpc.service="oteldemo.ShippingService"]
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:294]:
```suggestion
            rpc.method="ShipOrder" rpc.service="oteldemo.ShippingService"]
```

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+    name: "The product was shipped"
+    assertions:
+    - attr:rpc.grpc.status_code = 0
+    - attr:selected_spans.count >= 1
+  - selector: span[name="oteldemo.CartService/EmptyCart" rpc.system="grpc"
+			rpc.method="EmptyCart" rpc.service="oteldemo.CartService"]
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:300]:
```suggestion
            rpc.method="EmptyCart" rpc.service="oteldemo.CartService"]
```

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+    assertions:
+    - attr:rpc.grpc.status_code = 0
+    - attr:selected_spans.count >= 1
+```
+
+Finally, when running these tests we have the following report, showing each
+test file run in the transaction and also with the "Checkout” step described
+above:
+
+![results from the trace-based test](./trace-based-test-result.png)
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:311]: Since this is text only I recommend to make use of the text directly

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+
+```json
+{
+  "email": "google@example.com",
+  "order": {
+    "orderId": "505",
+    "shippingCost": {
+      "currencyCode": "USD",
+    },
+	  // ...
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:352]:
```suggestion
        // ...
```

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+      "currency_code": "USD",
+    },
+    // ...
+}
+```
+
+By doing that, we have a successful call to the service, and it evaluates
+correctly as seen here:
+
+![email-success-formatted.png](./email-success-formatted.png)
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:373]: Is there any reason why tracetest is used in this last screenshot and not jaeger or a console output as before? Also, the picture is really hard to parse on a small screen

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+
+As a team building an open source observability tool, we value the opportunity
+to contribute to the overall OpenTelemetry community. That's why we took action
+on [the issue](https://github.com/open-telemetry/opentelemetry-demo/issues/873)
+as soon as it was opened two months ago.
+
+Daniel Dias is working on Tracetest, an open source tool that allows you to
+develop and test your distributed system with OpenTelemetry. It works with any
+OTel-compatible system and enables trace–based tests to be created. Check it out
+at [https://github.com/kubeshop/tracetest](https://github.com/kubeshop/tracetest).
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:399]:
```suggestion
```

**Comment by svrnm** [comment]
thanks @adnanrahic , please see the comments I have provided

File: content/en/blog/2023/testing-otel-demo/index.md (hunk 4 above)
Comment (adnanrahic) [reply content/en/blog/2023/testing-otel-demo/index.md:65]: If we cannot have it as an image description, we can just drop this sentence.

File: content/en/blog/2023/testing-otel-demo/index.md (hunk 12 above)
Comment (adnanrahic) [reply content/en/blog/2023/testing-otel-demo/index.md:311]: Agreed. Changing to text.

File: content/en/blog/2023/testing-otel-demo/index.md (hunk 14 above)
Comment (adnanrahic) [reply content/en/blog/2023/testing-otel-demo/index.md:373]: The reasoning is to show the test specs and that the test is passing. I can try re-taking the screenshot and make it bigger.

File: content/en/blog/2023/testing-otel-demo/index.md (hunk 4 above)
Comment (adnanrahic) [reply content/en/blog/2023/testing-otel-demo/index.md]: I've removed the image description entirely.

**Comment by adnanrahic** [comment]
Thanks @svrnm! I've applied your suggestions.

I'll re-take the final screenshot to make the screen bigger and more readable.

Is there anything else I missed?

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 16]
```diff
@@ -0,0 +1,416 @@
...
+Since this test is a sequence of smaller tests, we created
+[a transaction](https://github.com/open-telemetry/opentelemetry-demo/blob/main/test/tracetesting/frontend-service/all.yaml)
+that defines the tests that will be run:
+
+```yaml
+type: Transaction
+spec:
+  name: 'Frontend Service'
+  description: Run all Frontend tests enabled in sequence,
+		simulating a process of a user purchasing products on the Astronomy store
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:194]:
```suggestion
        simulating a process of a user purchasing products on the Astronomy store
```

**Comment by adnanrahic** [comment]
Hi @svrnm 👋 

I've addressed all suggestions and made sure to properly format the post.
I'm happy to make additional edits if needed.

Thanks. 👌

**Comment by svrnm** [comment]
@adnanrahic thanks! I'll need a few days to be able to do another round of reviews, apologies, I'll follow up with you as soon as I can.

**Review by cartermp — APPROVED** [review]
Overall I think the blog is in good shape.

**Review by chalin — COMMENTED** [review]
I'll make a copyedit pass if I can soon too (though probably after @svrnm :)).

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 17]
```diff
@@ -0,0 +1,421 @@
...
+
+This term was popularized by Ted Young in the talk
+[Trace Driven Development: Unifying Testing and Observability](https://www.youtube.com/watch?v=NU-fTr-udZg)
+at KubeCon North America 2018.
+
+To perform a trace-based test, we execute an operation against the system, which
+generates a trace, following these steps:
+
+1. Trigger an operation against the system and collect its output and trace id
+   generated from the operation;
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:50]:
```suggestion
   generated from the operation.
```

**Review by svrnm — COMMENTED** [review]
minor issues, left, overall lgtm

Let's publish this by the end of the week/early next week as we just had another blog post yesterday

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 18]
```diff
@@ -0,0 +1,421 @@
+---
+title: Testing the OpenTelemetry Demo with Trace-based Tests
```
[the remark this answers is not in the document]
Comment (adnanrahic) [reply content/en/blog/2023/testing-otel-demo/index.md:2]: Sounds a lot better actually! 👌

**Comment by adnanrahic** [comment]
Thanks everyone!

I've applied your last edits. It looks great!

Would be awesome to see it published this week. 🙌

File: content/en/blog/2023/testing-otel-demo/index.md [hunk 19]
```diff
@@ -0,0 +1,421 @@
+---
+title: Trace-based Testing the OpenTelemetry Demo
+linkTitle: Testing the OTel Demo
+date: 2023-07-18
```
Comment (svrnm) [inline content/en/blog/2023/testing-otel-demo/index.md:4]:
```suggestion
date: 2023-07-27
```

**Comment by svrnm** [comment]
Let's publish it on Thursday then

**Comment by adnanrahic** [comment]
Wonderful @svrnm ! 🙌 
Thanks everyone for the reviews. 🙇

**Comment by adnanrahic** [comment]
Hey @svrnm ! Just checking in to ask when you plan on merging this? Wanted to know so I can help promote it once it's published. 😄

**Review by svrnm — APPROVED** [review]

**Comment by chalin** [comment]
Hi @adnanrahic - I've just rebased. Can you pull the rebased PR and then run `npm run fix:dict` and submit the changes? That'll normalize your local cSpell words. Thanks!

**Comment by svrnm** [comment]
published! https://opentelemetry.io/blog/2023/testing-otel-demo/

**Comment by adnanrahic** [comment]
Thanks a lot, everyone! 🙌
